// roblox-ts/out/TSTransformer/nodes/jsx/transformJsxChildren.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformJsxChildren = void 0;
const byots_1 = __importDefault(require("byots"));
const LuauAST_1 = __importDefault(require("../../../LuauAST"));
const diagnostics_1 = require("../../../Shared/diagnostics");
const assert_1 = require("../../../Shared/util/assert");
const DiagnosticService_1 = require("../../classes/DiagnosticService");
const transformExpression_1 = require("../expressions/transformExpression");
const createTypeCheck_1 = require("../../util/createTypeCheck");
const getKeyAttributeInitializer_1 = require("../../util/jsx/getKeyAttributeInitializer");
const offset_1 = require("../../util/offset");
const pointer_1 = require("../../util/pointer");
const types_1 = require("../../util/types");
function createJsxAddNumericChild(id, lengthId, key, value) {
    return LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.Assignment, {
        left: LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.ComputedIndexExpression, {
            expression: id,
            index: LuauAST_1.default.binary(lengthId, "+", key),
        }),
        operator: "=",
        right: value,
    });
}
function createJsxAddKeyChild(id, keyId, valueId) {
    return LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.Assignment, {
        left: LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.ComputedIndexExpression, {
            expression: id,
            index: keyId,
        }),
        operator: "=",
        right: valueId,
    });
}
function createJsxAddAmbiguousChildren(id, amtSinceUpdate, lengthId, expression) {
    const keyId = LuauAST_1.default.tempId("k");
    const valueId = LuauAST_1.default.tempId("v");
    return LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.ForStatement, {
        ids: LuauAST_1.default.list.make(keyId, valueId),
        expression: LuauAST_1.default.call(LuauAST_1.default.globals.pairs, [expression]),
        statements: LuauAST_1.default.list.make(LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.IfStatement, {
            condition: createTypeCheck_1.createTypeCheck(keyId, LuauAST_1.default.strings.number),
            statements: LuauAST_1.default.list.make(createJsxAddNumericChild(id, offset_1.offset(lengthId, amtSinceUpdate), keyId, valueId)),
            elseBody: LuauAST_1.default.list.make(createJsxAddKeyChild(id, keyId, valueId)),
        })),
    });
}
function createJsxAddArrayChildren(id, amtSinceUpdate, lengthId, expression) {
    const keyId = LuauAST_1.default.tempId("k");
    const valueId = LuauAST_1.default.tempId("v");
    return LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.ForStatement, {
        ids: LuauAST_1.default.list.make(keyId, valueId),
        expression: LuauAST_1.default.call(LuauAST_1.default.globals.ipairs, [expression]),
        statements: LuauAST_1.default.list.make(createJsxAddNumericChild(id, offset_1.offset(lengthId, amtSinceUpdate), keyId, valueId)),
    });
}
function createJsxAddMapChildren(id, expression) {
    const keyId = LuauAST_1.default.tempId("k");
    const valueId = LuauAST_1.default.tempId("v");
    return LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.ForStatement, {
        ids: LuauAST_1.default.list.make(keyId, valueId),
        expression: LuauAST_1.default.call(LuauAST_1.default.globals.pairs, [expression]),
        statements: LuauAST_1.default.list.make(createJsxAddKeyChild(id, keyId, valueId)),
    });
}
function countCreateJsxAddChildExpressionUses(isPossiblyUndefinedOrFalse, isPossiblyTrue, isPossiblyElement, isPossiblyArray, isPossiblyMap) {
    let expUses = 0;
    if (isPossiblyElement) {
        expUses += 1;
    }
    if (isPossiblyArray || isPossiblyMap) {
        expUses += 1;
        if (isPossiblyElement) {
            expUses += 3;
        }
    }
    if ((isPossiblyUndefinedOrFalse || isPossiblyTrue) && (isPossiblyElement || isPossiblyArray || isPossiblyMap)) {
        expUses += 1;
    }
    return expUses;
}
function createJsxAddChild(state, id, amtSinceUpdate, lengthId, expression, type) {
    const isPossiblyUndefinedOrFalse = types_1.isPossiblyType(type, t => types_1.isUndefinedType(t) || types_1.isBooleanLiteralType(state, t, false));
    const isPossiblyTrue = types_1.isPossiblyType(type, t => types_1.isBooleanLiteralType(state, t, true));
    const isPossiblyElement = types_1.isPossiblyType(type, t => types_1.isRoactElementType(state, t));
    let areMapKeysNonNumber = false;
    const isPossiblyMap = types_1.isPossiblyType(type, t => {
        if (types_1.isMapType(state, t)) {
            if (!areMapKeysNonNumber) {
                const typeArguments = types_1.getTypeArguments(state, t);
                if (types_1.isDefinitelyType(typeArguments[0], t => !types_1.isNumberType(t))) {
                    areMapKeysNonNumber = true;
                }
            }
            return true;
        }
        return false;
    });
    const isPossiblyArray = !areMapKeysNonNumber || types_1.isPossiblyType(type, t => types_1.isArrayType(state, t));
    const expUses = countCreateJsxAddChildExpressionUses(isPossiblyUndefinedOrFalse, isPossiblyTrue, isPossiblyElement, isPossiblyArray, isPossiblyMap);
    if (expUses > 1) {
        expression = state.pushToVarIfNonId(expression, "exp");
    }
    let statement;
    if (isPossiblyElement) {
        statement = createJsxAddNumericChild(id, lengthId, LuauAST_1.default.number(amtSinceUpdate + 1), expression);
    }
    if (isPossiblyArray || isPossiblyMap) {
        let loop;
        if (isPossiblyArray && isPossiblyMap) {
            loop = createJsxAddAmbiguousChildren(id, amtSinceUpdate, lengthId, expression);
        }
        else if (isPossiblyArray) {
            loop = createJsxAddArrayChildren(id, amtSinceUpdate, lengthId, expression);
        }
        else {
            loop = createJsxAddMapChildren(id, expression);
        }
        if (isPossiblyElement) {
            assert_1.assert(LuauAST_1.default.isAnyIdentifier(expression));
            const isFragmentCheck = LuauAST_1.default.binary(LuauAST_1.default.property(expression, "elements"), "~=", LuauAST_1.default.nil());
            const hasPropsCheck = LuauAST_1.default.binary(LuauAST_1.default.property(expression, "props"), "~=", LuauAST_1.default.nil());
            const hasComponentCheck = LuauAST_1.default.binary(LuauAST_1.default.property(expression, "component"), "~=", LuauAST_1.default.nil());
            const isElementCheck = LuauAST_1.default.binary(hasPropsCheck, "and", hasComponentCheck);
            const isElementLikeCheck = LuauAST_1.default.binary(isFragmentCheck, "or", isElementCheck);
            statement = LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.IfStatement, {
                condition: isElementLikeCheck,
                statements: LuauAST_1.default.list.make(statement),
                elseBody: LuauAST_1.default.list.make(loop),
            });
        }
        else {
            statement = loop;
        }
    }
    if (isPossiblyUndefinedOrFalse || isPossiblyTrue) {
        if (isPossiblyElement || isPossiblyArray || isPossiblyMap) {
            let condition;
            if (isPossiblyTrue) {
                condition = createTypeCheck_1.createTypeCheck(expression, LuauAST_1.default.strings.table);
            }
            else {
                condition = expression;
            }
            statement = LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.IfStatement, {
                condition,
                statements: LuauAST_1.default.list.make(statement),
                elseBody: LuauAST_1.default.list.make(),
            });
        }
        else {
            statement = LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.VariableDeclaration, {
                left: LuauAST_1.default.emptyId(),
                right: expression,
            });
        }
    }
    assert_1.assert(statement);
    return statement;
}
function transformJsxChildren(state, children, attributesPtr, childrenPtr) {
    const lengthId = LuauAST_1.default.tempId("length");
    let lengthInitialized = false;
    let amtSinceUpdate = 0;
    function updateLengthId() {
        const right = LuauAST_1.default.unary("#", childrenPtr.value);
        if (lengthInitialized) {
            state.prereq(LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.Assignment, {
                left: lengthId,
                operator: "=",
                right,
            }));
        }
        else {
            state.prereq(LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.VariableDeclaration, {
                left: lengthId,
                right,
            }));
            lengthInitialized = true;
        }
        amtSinceUpdate = 0;
    }
    function disableInline() {
        if (LuauAST_1.default.isMixedTable(childrenPtr.value)) {
            if (LuauAST_1.default.isMap(attributesPtr.value) && !LuauAST_1.default.list.isEmpty(attributesPtr.value.fields)) {
                pointer_1.disableMapInline(state, attributesPtr);
            }
            pointer_1.disableMixedTableInline(state, childrenPtr);
            updateLengthId();
        }
    }
    let lastUsefulElementIndex;
    for (lastUsefulElementIndex = children.length - 1; lastUsefulElementIndex >= 0; lastUsefulElementIndex--) {
        const child = children[lastUsefulElementIndex];
        if (!byots_1.default.isJsxText(child) || !child.containsOnlyTriviaWhiteSpaces)
            break;
    }
    for (let i = 0; i < children.length; i++) {
        const child = children[i];
        if (byots_1.default.isJsxText(child)) {
            if (!child.containsOnlyTriviaWhiteSpaces && child.text.match(/\S/)) {
                DiagnosticService_1.DiagnosticService.addDiagnostic(diagnostics_1.errors.noJsxText(child));
            }
            continue;
        }
        if (byots_1.default.isJsxExpression(child)) {
            const innerExp = child.expression;
            if (innerExp) {
                const [expression, prereqs] = state.capture(() => transformExpression_1.transformExpression(state, innerExp));
                if (!LuauAST_1.default.list.isEmpty(prereqs)) {
                    state.prereqList(prereqs);
                    disableInline();
                }
                if (child.dotDotDotToken) {
                    disableInline();
                    assert_1.assert(LuauAST_1.default.isAnyIdentifier(childrenPtr.value));
                    state.prereqList(prereqs);
                    state.prereq(createJsxAddArrayChildren(childrenPtr.value, amtSinceUpdate, lengthId, expression));
                }
                else {
                    const type = state.getType(innerExp);
                    if (types_1.isDefinitelyType(type, t => types_1.isRoactElementType(state, t))) {
                        if (LuauAST_1.default.isMixedTable(childrenPtr.value)) {
                            LuauAST_1.default.list.push(childrenPtr.value.fields, expression);
                        }
                        else {
                            state.prereq(createJsxAddNumericChild(childrenPtr.value, lengthId, LuauAST_1.default.number(amtSinceUpdate + 1), expression));
                        }
                        amtSinceUpdate++;
                    }
                    else {
                        disableInline();
                        assert_1.assert(LuauAST_1.default.isAnyIdentifier(childrenPtr.value));
                        state.prereq(createJsxAddChild(state, childrenPtr.value, amtSinceUpdate, lengthId, expression, type));
                    }
                }
                if (!LuauAST_1.default.isMixedTable(childrenPtr.value) && i < lastUsefulElementIndex) {
                    updateLengthId();
                }
            }
        }
        else {
            const [expression, prereqs] = state.capture(() => transformExpression_1.transformExpression(state, child));
            if (!LuauAST_1.default.list.isEmpty(prereqs)) {
                disableInline();
            }
            state.prereqList(prereqs);
            const keyInitializer = !byots_1.default.isJsxFragment(child) && getKeyAttributeInitializer_1.getKeyAttributeInitializer(child);
            if (keyInitializer) {
                const [key, keyPrereqs] = state.capture(() => transformExpression_1.transformExpression(state, keyInitializer));
                if (!LuauAST_1.default.list.isEmpty(keyPrereqs)) {
                    disableInline();
                }
                state.prereqList(keyPrereqs);
                pointer_1.assignToMixedTablePointer(state, childrenPtr, key, expression);
            }
            else {
                if (LuauAST_1.default.isMixedTable(childrenPtr.value)) {
                    LuauAST_1.default.list.push(childrenPtr.value.fields, expression);
                }
                else {
                    state.prereq(createJsxAddNumericChild(childrenPtr.value, lengthId, LuauAST_1.default.number(amtSinceUpdate + 1), expression));
                }
                amtSinceUpdate++;
            }
        }
    }
}
exports.transformJsxChildren = transformJsxChildren;
