// roblox-ts/out/TSTransformer/nodes/transformInitializer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformInitializer = void 0;
const LuauAST_1 = __importDefault(require("../../LuauAST"));
const transformExpression_1 = require("./expressions/transformExpression");
function transformInitializer(state, id, initializer) {
    return LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.IfStatement, {
        condition: LuauAST_1.default.binary(id, "==", LuauAST_1.default.nil()),
        elseBody: LuauAST_1.default.list.make(),
        statements: state.capturePrereqs(() => {
            state.prereq(LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.Assignment, {
                left: id,
                operator: "=",
                right: transformExpression_1.transformExpression(state, initializer),
            }));
        }),
    });
}
exports.transformInitializer = transformInitializer;
