// roblox-ts/out/TSTransformer/nodes/class/transformPropertyDeclaration.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformPropertyDeclaration = void 0;
const byots_1 = __importDefault(require("byots"));
const LuauAST_1 = __importDefault(require("../../../LuauAST"));
const diagnostics_1 = require("../../../Shared/diagnostics");
const DiagnosticService_1 = require("../../classes/DiagnosticService");
const transformExpression_1 = require("../expressions/transformExpression");
const transformObjectKey_1 = require("../transformObjectKey");
function transformPropertyDeclaration(state, node, ptr) {
    if (!byots_1.default.hasStaticModifier(node)) {
        return LuauAST_1.default.list.make();
    }
    if (byots_1.default.isPrivateIdentifier(node.name)) {
        DiagnosticService_1.DiagnosticService.addDiagnostic(diagnostics_1.errors.noPrivateIdentifier(node));
        return LuauAST_1.default.list.make();
    }
    if (!node.initializer) {
        return LuauAST_1.default.list.make();
    }
    return LuauAST_1.default.list.make(LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.Assignment, {
        left: LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.ComputedIndexExpression, {
            expression: ptr.value,
            index: transformObjectKey_1.transformObjectKey(state, node.name),
        }),
        operator: "=",
        right: transformExpression_1.transformExpression(state, node.initializer),
    }));
}
exports.transformPropertyDeclaration = transformPropertyDeclaration;
