// roblox-ts/out/Shared/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectType = exports.PARENT_FIELD = exports.MODULE_SUBEXT = exports.CLIENT_SUBEXT = exports.SERVER_SUBEXT = exports.INIT_NAME = exports.INDEX_NAME = exports.JSON_EXT = exports.LUA_EXT = exports.DTS_EXT = exports.D_EXT = exports.TSX_EXT = exports.TS_EXT = exports.RBXTS_SCOPE = exports.NODE_MODULES = exports.COMPILER_VERSION = exports.LIB_PATH = exports.PACKAGE_ROOT = void 0;
const path_1 = __importDefault(require("path"));
exports.PACKAGE_ROOT = path_1.default.join(__dirname, "..", "..");
exports.LIB_PATH = path_1.default.join(exports.PACKAGE_ROOT, "lib");
exports.COMPILER_VERSION = require("../../package.json").version;
exports.NODE_MODULES = "node_modules";
exports.RBXTS_SCOPE = "@rbxts";
exports.TS_EXT = ".ts";
exports.TSX_EXT = ".tsx";
exports.D_EXT = ".d";
exports.DTS_EXT = exports.D_EXT + exports.TS_EXT;
exports.LUA_EXT = ".lua";
exports.JSON_EXT = ".json";
exports.INDEX_NAME = "index";
exports.INIT_NAME = "init";
exports.SERVER_SUBEXT = ".server";
exports.CLIENT_SUBEXT = ".client";
exports.MODULE_SUBEXT = "";
exports.PARENT_FIELD = "Parent";
var ProjectType;
(function (ProjectType) {
    ProjectType["Game"] = "game";
    ProjectType["Model"] = "model";
    ProjectType["Package"] = "package";
})(ProjectType = exports.ProjectType || (exports.ProjectType = {}));
