// roblox-ts/out/Shared/diagnostics.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.warnings = exports.errors = exports.getDiagnosticId = void 0;
const byots_1 = __importDefault(require("byots"));
const kleur_1 = __importDefault(require("kleur"));
const createDiagnosticWithLocation_1 = require("./util/createDiagnosticWithLocation");
const createTextDiagnostic_1 = require("./util/createTextDiagnostic");
const REPO_URL = "https://github.com/roblox-ts/roblox-ts";
function suggestion(text) {
    return "Suggestion: " + kleur_1.default.yellow(text);
}
function issue(id) {
    return "More information: " + kleur_1.default.grey(`${REPO_URL}/issues/${id}`);
}
let id = 0;
function diagnostic(category, ...messages) {
    return diagnosticWithContext(category, undefined, ...messages);
}
function diagnosticWithContext(category, contextFormatter, ...messages) {
    const result = (node, context) => {
        if (category === byots_1.default.DiagnosticCategory.Error) {
            debugger;
        }
        if (contextFormatter) {
            messages.push(...contextFormatter(context));
        }
        return createDiagnosticWithLocation_1.createDiagnosticWithLocation(result.id, messages.join("\n"), category, node);
    };
    result.id = id++;
    return result;
}
function diagnosticText(category, ...messages) {
    return createTextDiagnostic_1.createTextDiagnostic(messages.join("\n"), category);
}
function error(...messages) {
    return diagnostic(byots_1.default.DiagnosticCategory.Error, ...messages);
}
function errorWithContext(contextFormatter, ...messages) {
    return diagnosticWithContext(byots_1.default.DiagnosticCategory.Error, contextFormatter, ...messages);
}
function warning(...messages) {
    return diagnostic(byots_1.default.DiagnosticCategory.Warning, ...messages);
}
function warningText(...messages) {
    return diagnosticText(byots_1.default.DiagnosticCategory.Warning, ...messages);
}
function getDiagnosticId(diagnostic) {
    return diagnostic.id;
}
exports.getDiagnosticId = getDiagnosticId;
exports.errors = {
    noInvalidIdentifier: error("Invalid Luau identifier!", "Luau identifiers must start with a letter and only contain letters, numbers, and underscores.", "Reserved Luau keywords cannot be used as identifiers."),
    noReservedIdentifier: error("Cannot use identifier reserved for compiler internal usage."),
    noReservedClassFields: error("Cannot use class field reserved for compiler internal usage."),
    noClassMetamethods: error("Metamethods cannot be used in class definitions!"),
    noForInStatement: error("for-in loop statements are not supported!"),
    noLabeledStatement: error("labels are not supported!"),
    noDebuggerStatement: error("`debugger` is not supported!"),
    noNullLiteral: error("`null` is not supported!", suggestion("Use `undefined` instead.")),
    noPrivateIdentifier: error("Private identifiers are not supported!"),
    noTypeOfExpression: error("`typeof` operator is not supported!", suggestion("Use `typeIs(value, type)` or `typeOf(value)` instead.")),
    noRegex: error("Regular expressions are not supported!"),
    noBigInt: error("BigInt literals are not supported!"),
    noAny: error("Using values of type `any` is not supported!", suggestion("Use `unknown` instead.")),
    noVar: error("`var` keyword is not supported!", suggestion("Use `let` or `const` instead.")),
    noGetterSetter: error("Getters and Setters are not supported!", issue(457)),
    noEqualsEquals: error("operator `==` is not supported!", suggestion("Use `===` instead.")),
    noExclamationEquals: error("operator `!=` is not supported!", suggestion("Use `!==` instead.")),
    noComma: error("operator `,` is not supported!"),
    noEnumMerging: error("Enum merging is not supported!"),
    noNamespaceMerging: error("Namespace merging is not supported!"),
    noSpreadDestructuring: error("Operator `...` is not supported for destructuring!"),
    noFunctionExpressionName: error("Function expression names are not supported!"),
    noPrecedingSpreadElement: error("Spread element must come last in a list of arguments!"),
    noDestructureAssignmentExpression: error("Cannot destructure LuaTuple<T> expression outside of an ExpressionStatement!"),
    noExportAssignmentLet: error("Cannot use `export =` on a `let` variable!", suggestion("Use `const` instead.")),
    noGlobalThis: error("`globalThis` is not supported!"),
    noArguments: error("`arguments` is not supported!"),
    noPrototype: error("`prototype` is not supported!"),
    noSuperProperty: error("super properties are not supported!"),
    noNonNumberStringRelationOperator: error("Relation operators can only be used on number or string types!"),
    noInstanceMethodCollisions: error("Static methods cannot use the same name as instance methods!"),
    noStaticMethodCollisions: error("Instance methods cannot use the same name as static methods!"),
    noUnaryPlus: error("Unary `+` is not supported!", suggestion("Use `tonumber(x)` instead.")),
    noNonNumberUnaryMinus: error("Unary `-` is only supported for number types!"),
    noAwaitForOf: error("`await` is not supported in for-of loops!"),
    noComplexForOf: error("for-of loops do not support complex variable lists!", issue(1253)),
    noAsyncGeneratorFunctions: error("Async generator functions are not supported!"),
    noNonStringModuleSpecifier: error("Module specifiers must be a string literal."),
    noIterableIteration: error("Iterating on Iterable<T> is not supported! You must use a more specific type."),
    noForOfLuaTupleAssignment: error("for-of assignment with LuaTuple<T> is not supported!"),
    noOptionalMacroCall: error("Macro methods can not be optionally called!"),
    noMixedTypeCall: error("Attempted to call a function with mixed types! All definitions must either be a method or a callback."),
    noIndexWithoutCall: error("Cannot index a method without calling it!", suggestion("Use the form `() => a.b()` instead of `a.b`.")),
    noMacroWithoutCall: error("Cannot index a macro without calling it!", suggestion("Use the form `() => a.b()` instead of `a.b`.")),
    noObjectWithoutMethod: error("Cannot access `Object` without calling a function on it!"),
    noConstructorMacroWithoutNew: error("Cannot index a constructor macro without using the `new` operator!"),
    noMacroExtends: error("Cannot extend from a macro class!"),
    noMacroUnion: error("Macro cannot be applied to a union type!"),
    noMacroObjectSpread: error("Macro classes cannot be used in an object spread!", suggestion("Did you mean to use an array spread? `[ ...exp ]`")),
    noVarArgsMacroSpread: error("Macros which use variadric arguments do not support spread expressions!", issue(1149)),
    noNonModule: error("File does not have an import or export statement!", suggestion("Add `export {};` as the first line."), issue(1043)),
    noModuleSpecifierFile: error("Could not find file for import. Did you forget to `npm install`?"),
    noInvalidModule: error("You can only use npm scopes that are listed in your typeRoots."),
    noUnscopedModule: error("You cannot use modules directly under node_modules."),
    noRojoData: errorWithContext((path) => [
        `Could not find Rojo data. There is no $path in your Rojo config that covers ${path}`,
    ]),
    noNonModuleImport: error("Cannot import a non-ModuleScript!"),
    noIsolatedImport: error("Attempted to import a file inside of an isolated container from outside!"),
    noRoactInheritance: error("Composition is preferred over inheritance with Roact components.", "More info: https://reactjs.org/docs/composition-vs-inheritance.html"),
    noSuperPropertyCallRoactComponent: error("`super` is not supported inside Roact components!"),
    noSuperConstructorRoactComponent: error("`super(props)` must be the first statement of the constructor in a Roact component!"),
    noJsxText: error("JSX text is not supported!"),
};
exports.warnings = {
    truthyChange: (checksStr) => warning(`value will be checked against ${checksStr}`),
    stringOffsetChange: (text) => warning(`String macros no longer offset inputs: ${text}`),
    rojoPathInSrc: (partitionPath, suggestedPath) => warningText(`Invalid Rojo configuration. $path fields should be relative to out directory.`, suggestion(`Change the value of $path from "${partitionPath}" to "${suggestedPath}".`)),
    runtimeLibUsedInReplicatedFirst: warning("This statement would generate a call to the runtime library. The runtime library should not be used from ReplicatedFirst."),
};
