// roblox-ts/out/TSTransformer/util/assignment.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCompoundAssignmentExpression = exports.createCompoundAssignmentStatement = exports.createAssignmentExpression = exports.getSimpleAssignmentOperator = void 0;
const byots_1 = __importDefault(require("byots"));
const LuauAST_1 = __importDefault(require("../../LuauAST"));
const createBinaryFromOperator_1 = require("./createBinaryFromOperator");
const types_1 = require("./types");
const COMPOUND_OPERATOR_MAP = new Map([
    [byots_1.default.SyntaxKind.MinusEqualsToken, "-="],
    [byots_1.default.SyntaxKind.AsteriskEqualsToken, "*="],
    [byots_1.default.SyntaxKind.SlashEqualsToken, "/="],
    [byots_1.default.SyntaxKind.AsteriskAsteriskEqualsToken, "^="],
    [byots_1.default.SyntaxKind.PercentEqualsToken, "%="],
    [byots_1.default.SyntaxKind.PlusPlusToken, "+="],
    [byots_1.default.SyntaxKind.MinusMinusToken, "-="],
    [byots_1.default.SyntaxKind.EqualsToken, "="],
]);
function getSimpleAssignmentOperator(leftType, operatorKind, rightType) {
    if (operatorKind === byots_1.default.SyntaxKind.PlusEqualsToken) {
        return types_1.isDefinitelyType(leftType, t => types_1.isStringType(t)) || types_1.isDefinitelyType(rightType, t => types_1.isStringType(t))
            ? "..="
            : "+=";
    }
    return COMPOUND_OPERATOR_MAP.get(operatorKind);
}
exports.getSimpleAssignmentOperator = getSimpleAssignmentOperator;
function createAssignmentExpression(state, readable, operator, value) {
    state.prereq(LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.Assignment, {
        left: readable,
        operator,
        right: value,
    }));
    return readable;
}
exports.createAssignmentExpression = createAssignmentExpression;
function createCompoundAssignmentStatement(state, node, writable, writableType, readable, operator, value, valueType) {
    return LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.Assignment, {
        left: writable,
        operator: "=",
        right: createBinaryFromOperator_1.createBinaryFromOperator(state, node, readable, writableType, operator, value, valueType),
    });
}
exports.createCompoundAssignmentStatement = createCompoundAssignmentStatement;
function createCompoundAssignmentExpression(state, node, writable, writableType, readable, operator, value, valueType) {
    return createAssignmentExpression(state, writable, "=", createBinaryFromOperator_1.createBinaryFromOperator(state, node, readable, writableType, operator, value, valueType));
}
exports.createCompoundAssignmentExpression = createCompoundAssignmentExpression;
