// roblox-ts/out/Project/functions/getParsedCommandLine.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getParsedCommandLine = void 0;
const byots_1 = __importDefault(require("byots"));
const validateCompilerOptions_1 = require("./validateCompilerOptions");
const DiagnosticError_1 = require("../../Shared/errors/DiagnosticError");
const ProjectError_1 = require("../../Shared/errors/ProjectError");
function createParseConfigFileHost() {
    return {
        fileExists: byots_1.default.sys.fileExists,
        getCurrentDirectory: byots_1.default.sys.getCurrentDirectory,
        onUnRecoverableConfigFileDiagnostic: d => {
            throw new DiagnosticError_1.DiagnosticError([d]);
        },
        readDirectory: byots_1.default.sys.readDirectory,
        readFile: byots_1.default.sys.readFile,
        useCaseSensitiveFileNames: byots_1.default.sys.useCaseSensitiveFileNames,
    };
}
function getParsedCommandLine(data) {
    const parsedCommandLine = byots_1.default.getParsedCommandLineOfConfigFile(data.tsConfigPath, {}, createParseConfigFileHost());
    if (parsedCommandLine === undefined) {
        throw new ProjectError_1.ProjectError("Unable to load TS program!");
    }
    else if (parsedCommandLine.errors.length > 0) {
        throw new DiagnosticError_1.DiagnosticError(parsedCommandLine.errors);
    }
    if (globalThis.RBXTSC_DEV) {
        parsedCommandLine.options.incremental = false;
        parsedCommandLine.options.tsBuildInfoFile = undefined;
    }
    validateCompilerOptions_1.validateCompilerOptions(parsedCommandLine.options, data.nodeModulesPath);
    return parsedCommandLine;
}
exports.getParsedCommandLine = getParsedCommandLine;
