// rbxts-transformer-flamework/out/information/statements/viewClassDeclaration.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.viewClassDeclaration = void 0;
var path_1 = __importDefault(require("path"));
var factory_1 = require("../../util/factory");
function viewClassDeclaration(state, node) {
    var e_1, _a;
    var _b, _c;
    var symbolProvider = state.symbolProvider;
    var symbol = state.getSymbol(node);
    var internalId = state.getUid(node);
    if (!node.name || !symbol)
        return;
    var isExternal = false;
    var decorators = [];
    var flameworkDecorators = new Set([
        symbolProvider.flameworkFile.get("Service"),
        symbolProvider.flameworkFile.get("Controller"),
        (_b = symbolProvider.componentsFile) === null || _b === void 0 ? void 0 : _b.get("Component"),
    ]);
    if (node.decorators) {
        try {
            for (var _d = __values(node.decorators), _e = _d.next(); !_e.done; _e = _d.next()) {
                var decorator = _e.value;
                if (!factory_1.f.is.call(decorator.expression))
                    continue;
                var symbol_1 = state.getSymbol(decorator.expression.expression);
                if (!symbol_1)
                    continue;
                if (!((_c = symbol_1.declarations) === null || _c === void 0 ? void 0 : _c[0]))
                    continue;
                if (!factory_1.f.is.identifier(decorator.expression.expression))
                    continue;
                var name_1 = decorator.expression.expression.text;
                var isFlameworkDecorator = flameworkDecorators.has(symbol_1);
                if (symbol_1 !== symbolProvider.flameworkFile.get("External")) {
                    decorators.push({
                        type: "WithNodes",
                        declaration: symbol_1.declarations[0],
                        arguments: decorator.expression.arguments.map(function (x) { return x; }),
                        internalId: state.getUid(symbol_1.declarations[0]),
                        isFlameworkDecorator: isFlameworkDecorator,
                        name: name_1,
                        symbol: symbol_1,
                    });
                }
                else {
                    isExternal = true;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
            }
            finally { if (e_1) throw e_1.error; }
        }
    }
    if (decorators.length > 0) {
        var classInfo = {
            name: node.name.text,
            internalId: internalId,
            isExternal: isExternal,
            node: node,
            decorators: decorators,
            symbol: symbol,
        };
        state.classes.set(symbol, classInfo);
        if (!state.isGame && !state.buildInfo.getBuildClass(internalId)) {
            var filePath = state.pathTranslator.getOutputPath(state.getSourceFile(node).fileName);
            var relativePath = path_1.default.relative(state.currentDirectory, filePath);
            state.buildInfo.addBuildClass({
                filePath: relativePath,
                internalId: internalId,
                isExternal: isExternal,
                decorators: decorators.map(function (x) { return ({
                    internalId: x.internalId,
                    isFlameworkDecorator: x.isFlameworkDecorator,
                    name: x.name,
                }); }),
            });
        }
    }
    else {
        var buildClass = state.buildInfo.getBuildClass(internalId);
        if (buildClass) {
            state.classes.set(symbol, {
                internalId: internalId,
                node: node,
                symbol: symbol,
                name: node.name.text,
                isExternal: buildClass.isExternal,
                decorators: buildClass.decorators.map(function (x) { return ({
                    type: "Base",
                    internalId: x.internalId,
                    name: x.name,
                    isFlameworkDecorator: x.isFlameworkDecorator,
                }); }),
            });
        }
    }
}
exports.viewClassDeclaration = viewClassDeclaration;
